### ISSUE TITLE
Same Wav file have a total different beam scores when it is assigned to different batches 

### ISSUE THREAD

**Issue by yanghongjiazheng** [rwth-i6/returnn-experiments#26]
First of all, thanks for your great works.  And I have just encountered a problem.  I find that a same wav file will have total different  beam scores When it's assigned to different batches.    The output_layer is "output" and the beam size is 16.  The beam score  is -4.7 when there is only one audio in the batch , and the beam score is  -103 when there are 26 audios, etc.  Can you explain why did this happen .  Thanks a lot

**Comment by albertz** [comment]
Which config is that? [This one](https://github.com/rwth-i6/returnn-experiments/blob/master/2018-asr-attention/librispeech/full-setup-attention/returnn.config)?
Which Returnn version?
On which data is this trained? Can you reproduce that with our pretrained model? Can you share the audio? Or reproduce that on some Librispeech sequence?

In batched decoding, i.e. when multiple sequences are together in the search, the option `max_seq_len` of the RecLayer is treated a bit different. E.g. in the mentioned config above, we have `"max_seq_len": "max_len_from('base:encoder')"`. `max_len_from` will be the max length of the whole batch, i.e. not per individual sequence. I.e. the decoder search will potentially run for more decoder steps, until it will pick the best sequence. And for all finished sequences in a batch, we do length normalization, i.e. the score of finished sequences in a batch is normalized such that is comparable to non-finished sequences. This means effectively that for your specific sequence, you might have a high length norm factor in it. Unfortunately it's not so easy to get the un-normalized score.
You could maybe try to disable the length normalization, and see if the score is the same in that case. But the overall WER might be worse.

**Comment by yanghongjiazheng** [comment]
Yes, the config is that.  The returnn version is c28b6f4f3431495f0dfa0bc04d8753cba4a81950. The training data contains Librispeech and other English speaking corpus. Because that I have changed some details of the config, I haven't try this with your pretrained model.   The audio last 4.93s. And the minmun length of audios in the batches is 0.84s.    I‘m wondering if the padding 0.0 will influence the search result ?

**Comment by albertz** [comment]
You might want to try a newer Returnn version also.
As said, I assume this is due to the length normalization, which will get different final scores depending on how big the batch is. I explained how you can calculate one final score into another.
Or you could maybe try to disable the length normalization, and see if the score is the same in that case. But the overall WER might be worse.

**Comment by yanghongjiazheng** [comment]
Thanks.  But how can I disable the length normalization

**Comment by albertz** [comment]
There is the flag `length_normalization` in the `ChoiceLayer`, which you can simply set to `False`.
In the config, that looks sth like this:
```
...
'output': {'class': 'choice', ..., 'length_normalization': False, ...},
...
```

**Comment by yanghongjiazheng** [comment]
Besides, the longest audio in the batch lasts 6.89s, the audio that I interest lasts 4.93s.  I calculate the score multiply the factor, the result should be -6.5  rather than -103

**Comment by albertz** [comment]
> Besides, the longest audio in the batch lasts 6.89s, the audio that I interest lasts 4.93s. I calculate the score multiply the factor, the result should be -6.5 rather than -103

You should look at the longest decoded label length, not the audio length.
E.g. if the longest label length is 100, and your specific label length is 5, you would get the factor 100/5, which is close to what you observe.

**Comment by yanghongjiazheng** [comment]
Thanks  a lot.  I have disabled the length_normalization function.  And that do work , the difference get closer.  However, there is still a small gap.  Actually , I have try the google mask schedule referring to `SpecAugment: A Simple Data Augmentation Method for Automatic Speech Recognition` .  Have your team try this augmentation?  If so , how it works.   I tried this, and it bring 2.66% WER decline in test-other. And the mask method will mask the part mfccs to zero which is same as the padding.  So I think this gap may come from this. I'll do several experiments to find it out .  Thanks again for your help!

**Comment by albertz** [comment]
> Thanks a lot. I have disabled the length_normalization function. And that do work , the difference get closer. However, there is still a small gap.


> Actually , I have try the google mask schedule referring to `SpecAugment: A Simple Data Augmentation Method for Automatic Speech Recognition` . Have your team try this augmentation? If so , how it works. I tried this, and it bring 2.66% WER decline in test-other. And the mask method will mask the part mfccs to zero which is same as the padding. So I think this gap may come from this. I'll do several experiments to find it out . Thanks again for your help!

Do you use this also during decoding? You likely have some randomness in it, so this might just be due to different random seed?

Yes, we also use SpecAugment. E.g. see [here](https://github.com/rwth-i6/returnn-experiments/tree/master/2019-asr-e2e-trafo-vs-lstm), or [here](https://github.com/rwth-i6/returnn-experiments/search?utf8=%E2%9C%93&q=random_mask&type=), or [here](https://github.com/rwth-i6/returnn-experiments/tree/master/2019-librispeech-system/attention).
